sdk/storage/Azure.Storage.Blobs.ChangeFeed/src/ChangeFeed.cs

```diff
@@ -0,0 +1,233 @@
+        /// </summary>
+        private readonly BlobContainerClient _containerClient;
+
+        /// <summary>
+        /// A <see cref="SegmentFactory"/> for creating new <see cref="Segment"/>s.
+        /// </summary>
+        private readonly SegmentFactory _segmentFactory;
+
+        /// <summary>
+        /// Queue of paths to years we haven't processed yet.
+        /// </summary>
+        private readonly Queue<string> _years;
+
+        /// <summary>
+        /// Paths to segments in the current year we haven't processed yet.
+        /// </summary>
+        private Queue<string> _segments;
+
+        /// <summary>
+        /// The Segment we are currently processing.
+        /// </summary>
+        private Segment _currentSegment;
+
+        /// <summary>
+        /// The latest time the Change Feed can safely be read from.
+        /// </summary>
+        private DateTimeOffset _lastConsumable;
+
+        /// <summary>
+        /// User-specified start time.  If the start time occurs before Change Feed was enabled
+        /// for this account, we will start at the beginning of the Change Feed.
+        /// </summary>
+        private DateTimeOffset? _startTime;
+
+        /// <summary>
+        /// User-specified end time.  If the end time occurs after _lastConsumable, we will
+        /// end at _lastConsumable.
+        /// </summary>
+        private DateTimeOffset? _endTime;
+
+        public ChangeFeed(
+            BlobContainerClient containerClient,
+            SegmentFactory segmentFactory,
+            Queue<string> years,
+            Queue<string> segments,
+            Segment currentSegment,
+            DateTimeOffset lastConsumable,
+            DateTimeOffset? startTime,
+            DateTimeOffset? endTime)
+        {
+            _containerClient = containerClient;
+            _segmentFactory = segmentFactory;
+            _years = years;
+            _segments = segments;
+            _currentSegment = currentSegment;
+            _lastConsumable = lastConsumable;
+            _startTime = startTime;
+            _endTime = endTime;
+        }
+
+        /// <summary>
+        /// Constructor for mocking, and for creating a Change Feed with no Events.
+        /// </summary>
+        public ChangeFeed() { }
+
+        // The last segment may still be adding chunks.
+        public async Task<Page<BlobChangeFeedEvent>> GetPage(
+            bool async,
+            int pageSize = Constants.ChangeFeed.DefaultPageSize,
+            CancellationToken cancellationToken = default)
+        {
+            if (!HasNext())
+            {
+                throw new InvalidOperationException("Change feed doesn't have any more events");
+            }
+
+            if (_currentSegment.DateTime > _endTime)
+            {
+                return new BlobChangeFeedEventPage();
+            }
+
+            if (_currentSegment.DateTime > _lastConsumable)
+            {
```

Comment:
Is this enough for `lastConsumable` checking?
I found the lastConsumable is not rounded to hour in the sample meta.json.
```json
{
    "version": 0,
    "lastConsumable": "2019-02-23T01:10:00.000Z",
    "storageDiagnostics": {
        "version": 0,
        "lastModifiedTime": "2019-02-23T02:24:00.556Z",
        "data": {
            "aid": "55e551e3-8006-0000-00da-ca346706bfe4",
            "lfz": "2019-02-22T19:10:00.000Z"
        }
    }
}
```
So if segment's  DateTime is 2019-02-23T01:00:00, then we may end up returning events later then lastConsumable?

Reply:
changed to _currentSegment.IsFinalized